params/hooks_libevm.go

```diff
@@ -51,33 +51,49 @@ func (RulesExtra) MinimumGasConsumption(x uint64) uint64 {
+	nativeasset.NativeAssetBalanceAddr: makePrecompile(&nativeasset.NativeAssetBalance{GasCost: AssetBalanceApricot}),
+	nativeasset.NativeAssetCallAddr:    makePrecompile(&nativeasset.NativeAssetCall{GasCost: AssetCallApricot, CallNewAccountGas: ethparams.CallNewAccountGas}),
 }
 
-var PrecompiledContractsApricotPhasePre6 = map[common.Address]contract.StatefulPrecompiledContract{
-	nativeasset.GenesisContractAddr:    &nativeasset.DeprecatedContract{},
-	nativeasset.NativeAssetBalanceAddr: &nativeasset.DeprecatedContract{},
-	nativeasset.NativeAssetCallAddr:    &nativeasset.DeprecatedContract{},
+var PrecompiledContractsApricotPhasePre6 = map[common.Address]vm.PrecompiledContract{
+	nativeasset.GenesisContractAddr:    makePrecompile(&nativeasset.DeprecatedContract{}),
+	nativeasset.NativeAssetBalanceAddr: makePrecompile(&nativeasset.DeprecatedContract{}),
+	nativeasset.NativeAssetCallAddr:    makePrecompile(&nativeasset.DeprecatedContract{}),
 }
 
-var PrecompiledContractsApricotPhase6 = map[common.Address]contract.StatefulPrecompiledContract{
-	nativeasset.GenesisContractAddr:    &nativeasset.DeprecatedContract{},
-	nativeasset.NativeAssetBalanceAddr: &nativeasset.NativeAssetBalance{GasCost: AssetBalanceApricot},
-	nativeasset.NativeAssetCallAddr:    &nativeasset.NativeAssetCall{GasCost: AssetCallApricot, CallNewAccountGas: ethparams.CallNewAccountGas},
+var PrecompiledContractsApricotPhase6 = map[common.Address]vm.PrecompiledContract{
+	nativeasset.GenesisContractAddr:    makePrecompile(&nativeasset.DeprecatedContract{}),
+	nativeasset.NativeAssetBalanceAddr: makePrecompile(&nativeasset.NativeAssetBalance{GasCost: AssetBalanceApricot}),
+	nativeasset.NativeAssetCallAddr:    makePrecompile(&nativeasset.NativeAssetCall{GasCost: AssetCallApricot, CallNewAccountGas: ethparams.CallNewAccountGas}),
 }
 
-var PrecompiledContractsBanff = map[common.Address]contract.StatefulPrecompiledContract{
-	nativeasset.GenesisContractAddr:    &nativeasset.DeprecatedContract{},
-	nativeasset.NativeAssetBalanceAddr: &nativeasset.DeprecatedContract{},
-	nativeasset.NativeAssetCallAddr:    &nativeasset.DeprecatedContract{},
+var PrecompiledContractsBanff = map[common.Address]vm.PrecompiledContract{
+	nativeasset.GenesisContractAddr:    makePrecompile(&nativeasset.DeprecatedContract{}),
+	nativeasset.NativeAssetBalanceAddr: makePrecompile(&nativeasset.DeprecatedContract{}),
+	nativeasset.NativeAssetCallAddr:    makePrecompile(&nativeasset.DeprecatedContract{}),
+}
+
+var PrecompiledContractsGranite = map[common.Address]vm.PrecompiledContract{
+	nativeasset.GenesisContractAddr:    makePrecompile(&nativeasset.DeprecatedContract{}),
+	nativeasset.NativeAssetBalanceAddr: makePrecompile(&nativeasset.DeprecatedContract{}),
+	nativeasset.NativeAssetCallAddr:    makePrecompile(&nativeasset.DeprecatedContract{}),
+}
+
+func init() {
+	for addr, precompile := range vm.PrecompiledContractsP256Verify {
```

Comment:
`vm.PrecompiledContractsP256Verify` has a comment "This is exported for testing purposes". Seems like we should either remove that comment in libevm, or define the address/precompile mapping ourselves here as the consumer of libevm.

Reply:
Would be curious what @ARR4N thinks when he's back next week. On one hand, defining the address in as few places as possible the better. On the other, since Geth does not register the precompile at a specific address upstream, it's a little odd for libevm to do so for all of its consumers. 

Either way, most importantly I think it just needs to be clear/well documented so that it is clear wherever it is defined that that is the source of truth used to register it by chains.

Reply:
> define the address/precompile mapping ourselves here as the consumer of libevm

I think this is the best option. When `libevm` doesn't absolutely need to change it shouldn't and, as @JonathanOppenheimer said, the address isn't going to change on C-Chain. 

> defining the address in as few places as possible the better

Agreed, but that should be kept in `coreth` IMO as `libevm` is nothing more than a provider of a VM in this instance. How it is configured (i.e. the precompile address) is the responsibility of the consumer.

Reply:
The `init` function will probably become unnecessary.